default.nix

```diff
@@ -130,12 +130,23 @@ let
 
       dalton = callPackage ./pkgs/apps/dalton { };
 
+      dirac = callPackage ./pkgs/apps/dirac rec {
+        stdenv = super.gcc8Stdenv;
```

Comment:
I don't think we need to override the c-compiler. Replacing gfortran is sufficient.

Reply:
ExaTensor explicitly states that it requires GCC 8 or 11 and it actually did not work with the normal `stdenv`

Reply:
> It might work then again to give the normal stdenv to Dirac?

I would try that first. All other dependencies are also still  built with the default `stdenv`. 

Also: does `exatensor` work with `gcc11stdenv`? In that case we can drop this hack as soon as upstream upgrades to gcc11.

Reply:
The ExaTensor stuff is unfortunately very hacky but required. There is a combination of the following problems, which make this the only viable option, i would guess:

- OpenMPI does not build with `gcc11stdenv` due to implicit typecasts from REAL(8) to COMPLEX(8), which were apparently OK in older versions. One could probably fiddle around with the compiler flags to make this work, but that might be worse than the current solution.
- ExaTensor uses `mpifort` to get the linking correct. I've tried with replacing the `mpifort` calls with `gfortran` (and gfortran coming from `gfortran8` or `gfortran11` but then it doesn't find the MPI libraries. Again, one could probably patch this until it works, but at which price.
- GFortran 9 and 10 actually throw internal compiler errors on those files, I guess that would not be an easy fix.

Reply:
This combination works just fine with minimal overrides.

```nix
      dirac = callPackage ./pkgs/apps/dirac rec {
        exatensor = self.exatensor;
      };
      exatensor = callPackage ./pkgs/apps/exatensor rec {
        mpi = super.mpi.override { gfortran = super.gfortran8; };
      };

```

Reply:
> This combination works just fine with minimal overrides.

That's probably the best solution.

> OpenMPI does not build with gcc11stdenv due to...

That's good to know. This will most likely affect many more packages than just openmpi(?).

> ExaTensor uses mpifort to get the linking correct. I've tried with replacing the mpifort calls with gfortran (and gfortran coming from gfortran8 or gfortran11 but then it doesn't find the MPI libraries.

The default compiler paths are hard coded into the `openmpi` installation. One can override the default compilers via environment variables such as `OMPI_FC`.

Reply:
> > This combination works just fine with minimal overrides.
> 
> That's probably the best solution.
> 
> > OpenMPI does not build with gcc11stdenv due to...
> 
> That's good to know. This will most likely affect many more packages than just openmpi(?).
> 
> > ExaTensor uses mpifort to get the linking correct. I've tried with replacing the mpifort calls with gfortran (and gfortran coming from gfortran8 or gfortran11 but then it doesn't find the MPI libraries.
> 
> The default compiler paths are hard coded into the `openmpi` installation. One can override the default compilers via environment variables such as `OMPI_FC`.

Ah that is good to know. I've changed them in the ExaTensor makefile to gfortran directly and this is obviously not a good solution ^^

Reply:
For reference with the compiler hard coding: https://github.com/NixOS/nixpkgs/blob/a6a026abe97faa42db6ccd48f3be0b8baf4cb26f/pkgs/development/libraries/openmpi/default.nix#L104

(MPICH and MVAPICH follow the same logic). 

You can in principal use gfortran directly if you add the right compiler flags, obtained from `mpifortran -showme:compile` or `mpifortran -showme:link`.


(The disadvantage of both solutions is that the keywords/variable names are specific to openmpi.)